README.md
# multiArity 
Overload functions by arity.


## Install

```bash
$ npm install --save multiarity
```


## Usage

```javascript
var multiarity = require('multiarity');

// *** FOR THE IMPATIENT ***

var unsafeMultiply = function (a, b) { return a * b; };

console.log( unsafeMultiply(1) ); //=> NAN. USELESS!!!!!

// Overloading for sensible defaults!!!

var betterMultiply = multiarity(
	function () { return this.recur(1,1); },
	function (a) { return this.recur(a, 1); },
	function (a, b) { return a * b; } );

console.log(betterMultiply()); //=> 1
console.log(betterMultiply(7)); //=> 7
console.log(betterMultiply(7, 2)); //=> 14


// Overloading for stack-blowing recursion!!!
var contrived_factorial = multiarity(
	function () { return 1;},
	function (n) {
		if (n > 0) {
			return n * this.recur(n - 1);
		}
		return this.recur();
	});

console.log(contrived_factorial(3));

// *** FOR THE PATIENT ***

/**
 * Lets make a function that greets things by name. 
 */
var sayHelloTo = function (name) {
	console.log("[line:17]  sayHelloTo was called!");
	console.log("[line:18]  hello", name);
};

sayHelloTo("sayHelloTo_GUY"); // LOGS: sayHelloTo was called! \n "hello sayHelloTo_GUY"

/**
 * But what if I don't specify a name?
 */


sayHelloTo(); // LOGS: sayHelloTo was called! \n hello undefined


/**
 * ^^That's probably not what we wanted.
 * We probably want to have some default
 * value for `name` if one isn't specified.
 */

var sayHelloWithDefault = function () {
	if (arguments.length > 0) {
		console.log("[line:44]  hello", arguments[0]);
	} else {
		console.log("[line:46]  hello world");
	}
};

sayHelloWithDefault(); //LOGS: "hello world"
sayHelloWithDefault("sayHelloWithDefault_GUY"); //LOGS: "hello sayHelloWithDefault_GUY"

/**
 * Thats better... but there's a bit of duplication in our code:
 * both branches of our if statement are calling `console.log`...
 * the only difference between them is the number of arguments.
 *
 * It would be better, IMO, if we could express our intent
 * more explicitly as well. So what is our intent?
 *
 * (1) Our primary intent is to make a function that takes a name
 * and prints "hello" + name.
 * 
 * (2) Our secondary intent is to make a function that returns "hello" + "world" when
 * no arguments are passed to it. 
 * 
 * Wait, lets try breaking that down further...
 * 	(2.a) make a function that returns "hello" + "world"
 * 	(2.b) use 2.a when no arguments are provided
 *
 *
 * Our origional implamentation of `sayHelloTo` will do nicely for intent 1.
 * 
 * We could even re-use it for 2.a:
 * 		`sayHelloTo("world")` will return "hello" + "world".
 * 
 * Now finaly we need something for 2.b. Perhaps, a function that delegates
 * to several other functions depending on the number of arguments?
 */


/**
 * Suspend your disbelief for a moment and imagine that we
 * had a function, `multiarity`, that could do such a thing...
 * we could make a function that calls sayHelloTo when
 * one argument is provided like so:
 */
var sayHelloSingleArity = multiarity({
	1: sayHelloTo // we use a key of 1 to say this is what to call when 1 argument is provided
});

sayHelloSingleArity("sayHelloSingleArity_GUY"); // LOGS: sayHelloTo was called! \n "hello sayHelloSingleArity_GUY"

try {
	sayHelloSingleArity(); // ERROR!
} catch(e) {
	console.log("[line:97] ",e);
}


/**
 * *** A MULTIPLE-DISPATCH version!!! ***
 * ok lets teach say `sayHelloSingleArity` how to handle
 * two arguments: 
 */

var sayHelloDoubleArity = multiarity({
	1: sayHelloTo,
	0: function() {
		/**
		 * Within any of the functions passed into the 
		 * multiarity combinator `this.recur` is a 
		 * refrence to the new function that the multiarity combinator
		 * will return.
		 *
		 * So: `this.recur("world") === sayHelloDoubleArity("world");`
		 */
		console.log("[line:118] sayHelloDoubleArity called with 0 arguments");
		this.recur("world");
	}
});

/**
 * This should make sense, one argument provided means we dispatch to sayHelloTo
 * which in turn prints `sayHelloTo was called` and then our greeting.
 */
sayHelloDoubleArity("sayHelloDoubleArity_GUY"); 
// LOGS: 	sayHelloTo was called!
// 				"sayHelloDoubleArity_GUY"

/**
 * And this makes sense too, the zero argument version is dispatched to,
 * which simply recals the function (a la this.recur) with 1 argument of "world".
 * The one argument call dispatches to sayHelloTo as it always does and 
 * the world is once again greeted by a computer.
 */
sayHelloDoubleArity(); 
// LOGS: 	sayHelloDoubleArity called with 0 arguments
// 				sayHelloTo was called! 
// 				hello world

/**
 * BINGO. And for power users, leave off the argument numbering and let 
 * multiarity infer it based on your params list for you.
 */

var sayHelloLikeAPro = multiarity(
	function(name){ console.log("[LINE:143] hello pro", name); },
	function(){ this.recur("world"); });


sayHelloLikeAPro(); //LOGS: hello pro world

sayHelloLikeAPro("grammer"); // :)


```

## API

_(Coming soon)_


## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using [gulp](http://gulpjs.com/).


## License

Copyright (c) 2014 Andrew Stern. Licensed under the MIT license.



[npm-url]: https://npmjs.org/package/multiarity
[npm-image]: https://badge.fury.io/js/multiarity.svg
[travis-url]: https://travis-ci.org/androidStern/multiarity
[travis-image]: https://travis-ci.org/androidStern/multiarity.svg?branch=master
[daviddm-url]: https://david-dm.org/androidStern/multiarity.svg?theme=shields.io
[daviddm-image]: https://david-dm.org/androidStern/multiarity
[coveralls-url]: https://coveralls.io/r/androidStern/multiarity
[coveralls-image]: https://coveralls.io/repos/androidStern/multiarity/badge.png


CHANGELOG.md

# 0.1.0 (2014-11-11)

- Initial version


test/multiarity_test.js
'use strict';
var _ = require("lodash");
var multiarity = require('../build/multiarity.js');
var expect = require('chai').expect;

describe('multiarity', function () {

  it('should be a function', function () {
      expect(multiarity).to.be.a("function");
  });

  it("should require at least 1 argument", function(){
    expect(multiarity).to.throw("multiarity expects at least 1 arument");
  });

  describe("when passed a map", function (){
    var invalidArityMap = function (opts) {
      return _.merge({
        0: _.noop
      }, opts);
    }

    var otherwise = function () {return { "otherwise": _.noop}; };

    var validArityMap = _.compose(invalidArityMap, otherwise);

    it("should throw if all vals of map aren't functions", function (){
      var doarity = function () {
        multiarity({1: "hello"});
      }
      expect(doarity).to.throw("multiarity vals must be type function");
    });

    it("should return a function", function(){
      expect(multiarity(validArityMap())).to.be.a("function");
    });

    it("should have its return function's contexts bound to recur", function(){
      var arity_map = _.extend(validArityMap(), {
        1: function () {
          expect(this).to.have.property("recur");
        }
      });
      var testFunc = multiarity(validArityMap());
      testFunc(1);
    });

    it("should accept an otherwise case", function () {
      var oneIsFalseElseTrue = multiarity({
        "1": function () {
          return false;
        },
        "otherwise": function () {
          return true;
        }
      });
      expect(oneIsFalseElseTrue(1)).to.be.false
      expect(oneIsFalseElseTrue()).to.be.true
      expect(oneIsFalseElseTrue(1,2,3)).to.be.true
    });

    it("should return undefined if an arity is not found and an otherwise case is not provided", function(){
      var onlyOneIsFalse = multiarity({
        1: function () {return false; }
      });
      expect(onlyOneIsFalse).to.throw("wrong number of args: 0");
    });

    it("this.recur should call the function recursivley [1]", function (){
      var add = multiarity({
        "1": function (a) {
          return this.recur(a, 0);
        },
        "2": function (a, b) {
          return a + b;
        }
      });
      expect(add(1)).to.eq(1);
    });

    it("this.recur should call the function recursivley [2]", function (){
      var sayHello = multiarity({
        0: function () {
          return this.recur("world");
        },
        1: function (name) {
          return "hello " + name;
        }
      });
      expect(sayHello()).to.eq("hello world");
      expect(sayHello("andrew")).to.eq("hello andrew");
    });  
  });

  describe("when called with a list of functions", function(){

    it("should infer arities based on function lengths", function(){
      var sayHello = multiarity(
        function(){
          return this.recur("world");
        },
        function(name) {
          return "hello " + name;
        }
      );
      expect(sayHello()).to.eq("hello world");
    });

  });
  
  
});


gulpfile.js
'use strict';

var gulp   = require('gulp');
var plugins = require('gulp-load-plugins')();
var noop = function(){};
var paths = {
  lint: ['./gulpfile.js'],
  watch: ['./gulpfile.js', './lib/**', './test/**/*.js', '!test/{temp,temp/**}'],
  tests: ['./test/**/*.js', '!test/{temp,temp/**}'],
  source: ['./lib/*.js'],
  dest: ['./build']
};

var plumberConf = {};

if (process.env.CI) {
  plumberConf.errorHandler = function(err) {
    throw err;
  };
}

gulp.task('build', function(cb) {
  gulp.src(paths.source)
    .pipe(plugins.sourcemaps.init())
    .pipe(plugins.sweetjs({
      modules: ['lambda-jam/macro']
    }))
    .pipe(plugins.sourcemaps.write('.'))
    .pipe(gulp.dest('build'))
    .on('end', cb)
    .on('error', noop);
});

gulp.task('lint', function () {
  return gulp.src(paths.lint)
    .pipe(plugins.jshint('.jshintrc'))
    .pipe(plugins.plumber(plumberConf))
    .pipe(plugins.jscs())
    .pipe(plugins.jshint.reporter('jshint-stylish'))
    .on('error', noop);
});

gulp.task('istanbul', ['build'], function (cb) {
  gulp.src(paths.source)
    .pipe(plugins.istanbul()) // Covering files
    .on('finish', function () {
      gulp.src(paths.tests)
        .pipe(plugins.plumber(plumberConf))
        .pipe(plugins.mocha())
        .pipe(plugins.istanbul.writeReports()) // Creating the reports after tests runned
        .on('finish', function() {
          process.chdir(__dirname);
          cb();
        })
        .on('error',noop);
    });
});

gulp.task('bump', ['build', 'test'], function () {
  var bumpType = plugins.util.env.type || 'patch'; // major.minor.patch

  return gulp.src(['./package.json'])
    .pipe(plugins.bump({ type: bumpType }))
    .pipe(gulp.dest('./'));
});

gulp.task('watch', ['build', 'test'], function () {
  gulp.watch(paths.watch, ['build', 'test']);
});

gulp.task('test', ['build', 'istanbul']);

gulp.task('release', ['bump']);

gulp.task('default', ['lint', 'watch']);


lib/multiarity.js
/*
 * multiArity
 * https://github.com/androidStern/multiarity
 *
 * Copyright (c) 2014 Andrew Stern
 * Licensed under the MIT license.
 */

'use strict';

var _ = require("lodash");

module.exports = function multiArity() {
	if (arguments.length === 0) {
		throw new Error("multiarity expects at least 1 arument");
	}
	
	var args = _.toArray(arguments),
			arities = {},
			ctx = { recur: callByArity },
			opts = _.first(args);
	
	if (_.isPlainObject(opts)) {
		if (!isAllFns(_.values(opts))) {
			throw new Error("multiarity vals must be type function");
		}
		_.extend(arities, opts);
	} else {
		_.extend(arities, arityMapForFunctions(args));
	}

  return callByArity;

  function callByArity () {
  	var args = _.toArray(arguments),
  			fn_len = args.length,
  			fn_for_arity = _.has(arities, fn_len) ? 
  				arities[fn_len] : arities["otherwise"];
				if (_.isUndefined(fn_for_arity)) {
					throw new Error("wrong number of args: " + fn_len);
				}
		return fn_for_arity.apply(ctx, args);
  }
};

function arityMapForFunctions (fns_list) {
	return _.zipObject(_.map(fns_list, λ(_1.length)), fns_list);
}

function isAllFns (arr) {
	return _.all(arr, _.isFunction);
}




example/simple.js
/*
 * multiArity
 * https://github.com/androidStern/multiarity
 *
 * Copyright (c) 2014 Andrew Stern
 * Licensed under the MIT license.
 */

var multiarity = require('../');

// *** FOR THE IMPATIENT ***

var unsafeMultiply = function (a, b) { return a * b; };

console.log( unsafeMultiply(1) ); //=> NAN. USELESS!!!!!

var betterMultiply = multiarity(
	function () { return this.recur(1,1); },
	function (a) { return this.recur(a, 1); },
	function (a, b) { return a * b; } );

console.log(betterMultiply()); //=> 1
console.log(betterMultiply(7)); //=> 7
console.log(betterMultiply(7, 2)); //=> 14


var contrived_factorial = multiarity(
	function () { return 1;},
	function (n) {
		if (n > 0) {
			return n * this.recur(n - 1);
		}
		return this.recur();
	});

console.log(contrived_factorial(3));

// *** FOR THE PATIENT ***

/**
 * Lets make a function that greets things by name. 
 */
var sayHelloTo = function (name) {
	console.log("[line:17]  sayHelloTo was called!");
	console.log("[line:18]  hello", name);
};

sayHelloTo("sayHelloTo_GUY"); // LOGS: sayHelloTo was called! \n "hello sayHelloTo_GUY"

/**
 * But what if I don't specify a name?
 */


sayHelloTo(); // LOGS: sayHelloTo was called! \n hello undefined


/**
 * ^^That's probably not what we wanted.
 * We probably want to have some default
 * value for `name` if one isn't specified.
 */

var sayHelloWithDefault = function () {
	if (arguments.length > 0) {
		console.log("[line:44]  hello", arguments[0]);
	} else {
		console.log("[line:46]  hello world");
	}
};

sayHelloWithDefault(); //LOGS: "hello world"
sayHelloWithDefault("sayHelloWithDefault_GUY"); //LOGS: "hello sayHelloWithDefault_GUY"

/**
 * Thats better... but there's a bit of duplication in our code:
 * both branches of our if statement are calling `console.log`...
 * the only difference between them is the number of arguments.
 *
 * It would be better, IMO, if we could express our intent
 * more explicitly as well. So what is our intent?
 *
 * (1) Our primary intent is to make a function that takes a name
 * and prints "hello" + name.
 * 
 * (2) Our secondary intent is to make a function that returns "hello" + "world" when
 * no arguments are passed to it. 
 * 
 * Wait, lets try breaking that down further...
 * 	(2.a) make a function that returns "hello" + "world"
 * 	(2.b) use 2.a when no arguments are provided
 *
 *
 * Our origional implamentation of `sayHelloTo` will do nicely for intent 1.
 * 
 * We could even re-use it for 2.a:
 * 		`sayHelloTo("world")` will return "hello" + "world".
 * 
 * Now finaly we need something for 2.b. Perhaps, a function that delegates
 * to several other functions depending on the number of arguments?
 */


/**
 * Suspend your disbelief for a moment and imagine that we
 * had a function, `multiarity`, that could do such a thing...
 * we could make a function that calls sayHelloTo when
 * one argument is provided like so:
 */
var sayHelloSingleArity = multiarity({
	1: sayHelloTo // we use a key of 1 to say this is what to call when 1 argument is provided
});

sayHelloSingleArity("sayHelloSingleArity_GUY"); // LOGS: sayHelloTo was called! \n "hello sayHelloSingleArity_GUY"

try {
	sayHelloSingleArity(); // ERROR!
} catch(e) {
	console.log("[line:97] ",e);
}


/**
 * *** A MULTIPLE-DISPATCH version!!! ***
 * ok lets teach say `sayHelloSingleArity` how to handle
 * two arguments: 
 */

var sayHelloDoubleArity = multiarity({
	1: sayHelloTo,
	0: function() {
		/**
		 * Within any of the functions passed into the 
		 * multiarity combinator `this.recur` is a 
		 * refrence to the new function that the multiarity combinator
		 * will return.
		 *
		 * So: `this.recur("world") === sayHelloDoubleArity("world");`
		 */
		console.log("[line:118] sayHelloDoubleArity called with 0 arguments");
		this.recur("world");
	}
});

/**
 * This should make sense, one argument provided means we dispatch to sayHelloTo
 * which in turn prints `sayHelloTo was called` and then our greeting.
 */
sayHelloDoubleArity("sayHelloDoubleArity_GUY"); 
// LOGS: 	sayHelloTo was called!
// 				"sayHelloDoubleArity_GUY"

/**
 * And this makes sense too, the zero argument version is dispatched to,
 * which simply recals the function (a la this.recur) with 1 argument of "world".
 * The one argument call dispatches to sayHelloTo as it always does and 
 * the world is once again greeted by a computer.
 */
sayHelloDoubleArity(); 
// LOGS: 	sayHelloDoubleArity called with 0 arguments
// 				sayHelloTo was called! 
// 				hello world

/**
 * BINGO. And for power users, leave off the argument numbering and let 
 * multiarity infer it based on your params list for you.
 */

var sayHelloLikeAPro = multiarity(
	function(name){ console.log("[LINE:143] hello pro", name); },
	function(){ this.recur("world"); });


sayHelloLikeAPro(); //LOGS: hello pro world

sayHelloLikeAPro("grammer"); // :)

